# GBS-based-population-structure-analysis-of-Ross-Sea-fish
setwd("/Users/sepva552/Desktop/GBS/CHW.60")

library(vcfR)
library(adegenet)
library(StAMPP)
library(ggplot2)
library(corrplot)
library(LEA)
library(pophelper)
library(reshape2)
library(diveRsity)
library(corrplot)
library(devtools)
library(diveRsity)
library(vegan)
library(SNPRelate)
library(gdsfmt)
library(ape)
library(phangorn)
library(parallel)
library(dartR)
library(pegas)

##############################
snp_vcf2 = read.vcfR("./final.recode.vcf")
pop.data2 = read.table("./CHW_pop.txt", header = F)
gl.snp2 <- vcfR2genlight(snp_vcf2)
pop(gl.snp2) <- rep(pop.data2$V2)
snp.pca2 <- glPca(gl.snp2, nf = 10)
snp.pca.scores2 <- as.data.frame(snp.pca2$scores)
snp.pca.scores2$pop <- pop(gl.snp2)
write.table(snp.pca.scores2, "./CHW_PCA.txt", sep = "\t")
eig.val<-snp.pca2$eig
eig.val
nInd(gl.snp2)
length(pop.data2$V2) 
#percentages of variance for each PC
eig.perc <- 100*snp.pca2$eig/sum(snp.pca2$eig)
eig.perc
eigen<-data.frame(eig.val,eig.perc)
eigen


write.csv(eigen,file="./Stereno_adegenetPCA_eigen-summary.csv",row.names=TRUE,quote=FALSE)


##PCA plotting
data2 = read.delim("CHW_adegenetPCA.txt") #I have manually added population information to this file prior to loading it
mycol = c("#f1c039", "#f37d21", "#51692d", "#56ba32", "#87CEFA")
ggplot(data2, aes(x=PC1, y=PC2, color=pop)) +
  geom_point(size = 3) + 
  scale_color_manual(values=mycol) +
  theme_classic()+
  xlab("PC1 (28.60%)") +
  ylab("PC2 (4.58%)")


##calculating Fst between populations
Fish_Fst <- stamppFst(gl.snp2, nboots = 100, percent = 95, nclusters = 4)
Fst <- Fish_Fst$Fsts
pFst <- Fish_Fst$Pvalues
write.table(Fst, "Fst1.txt", sep="\t")
write.table(pFst, "Fst1_pvalue.txt", sep="\t")
##creating heatmap
# Melt the correlation matrix
Fish_Fst <- as.matrix(read.table("Fst1.txt"))
FishFs <- melt(Q_Fst, na.rm = TRUE)
summary(QflyFs$value)
ggplot(data = FishFs, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "#ffd60a", high = "#001d3d", mid = "#4e9de6", 
                       midpoint = 0.056, limit = c(0.005,0.11), space = "Lab", 
                       name="Pairwise Fst") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1))+
  coord_fixed()


library(LEA)
library(pophelper)

##creating input files
vcf2geno(input.file = "final.recode.vcf", output.file = "Qff.geno")

##snmf clustering
projectalpha = NULL
projectalpha = snmf("Qff.geno", K = 1:10, repetitions = 50, entropy = T, CPU = 8, project = "new")

# plot cross-entropy criterion for all runs in the snmf project
pdf(file = "./cross_ent_alphadefualt.pdf")
plot(projectalpha, col = "maroon4", pch = 19, cex = 1.2)
dev.off()

best2 = which.min(cross.entropy(projectalpha, K = 2))
best2
best3 = which.min(cross.entropy(projectalpha, K = 3))
best3
best4 = which.min(cross.entropy(projectalpha, K = 4))
best4
best5 = which.min(cross.entropy(projectalpha, K = 5))
best5
best6 = which.min(cross.entropy(projectalpha, K = 6))
best6
best7 = which.min(cross.entropy(projectalpha, K = 7))
best7
best8 = which.min(cross.entropy(projectalpha, K = 8))
best8
best9 = which.min(cross.entropy(projectalpha, K = 9))
best9
best10 = which.min(cross.entropy(projectalpha, K = 10))
best10


##creating admixture plots. For this, you need to first create a new folder (All_Qfiles) and move the Q files with "best" entropies from the LEA runs into it. 
sfiles <- list.files(path=("./All_Qfiles"), full.names=T)
slist <- readQ(files=sfiles)
clustercol <- c("#51692d", "#f1c039", "#f37d21", "#56ba32", "#a1d99b",
                "#9ecae1", "#fc9272", "#dd1c77", "#756bb1", "#636363")
plotQ(qlist=slist[3],imgtype = "pdf",
      height = 1.5, clustercol = c("#51692d","#56ba32","#f1c039"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[4],imgtype = "pdf",
      height = 1.5, clustercol = c("#51692d","#56ba32","#f1c039","#f37d21"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[5],imgtype = "pdf",
      height = 1.5, clustercol = c("#f37d21","#51692d","#f1c039","#56ba32","#a63838"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[6],imgtype = "pdf",
      height = 1.5, clustercol = c("#a63838","#f1c039","#ecbcab","#56ba32","#51692d","#f37d21"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[7],imgtype = "pdf",
      height = 1.5, clustercol = c("#51692d","#a63838","#f1c039","#ecbcab","#caf291","#56ba32","#f37d21"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[8],imgtype = "pdf",
      height = 1.5, clustercol = c("#3d87db","#a63838","#51692d","#f1c039","#ecbcab","#caf291","#56ba32","#f37d21"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[9],imgtype = "pdf",
      height = 1.5, clustercol = c("#f1c039","#a63838","#3d87db","#ecbcab","#caf291","#56ba32","#0000FF","#51692d","#f37d21"), dpi = 1200, exportpath = "./")
plotQ(qlist=slist[1],imgtype = "pdf",
      height = 1.5, clustercol = c("#dd00ff","#51692d","#caf291","#3d87db","#ecbcab","#a63838","#56ba32","#0000FF","#f1c039","#f37d21"), dpi = 1200, exportpath = "./") 




##############AMOVA
library(vcfR)       # Read VCF
library(adegenet)   # Convert to genind
library(poppr)      # AMOVA
library(dartR)      # Quality checks (optional)
install.packages("dartR")
# Read VCF
vcf <- read.vcfR("final.recode.vcf")

# Read population assignments
pop_data <- read.table("Combo.txt", header = FALSE, col.names = c("Sample", "Population"))

# Convert VCF to genind
genind_data <- vcfR2genind(vcf)

# Assign populations (order must match VCF samples!)
pop(genind_data) <- pop_data$Population[match(indNames(genind_data), pop_data$Sample)]

# Verify
table(pop(genind_data))

genind_data <- genind_data[pop(genind_data) != "Pop1", ]
# Basic AMOVA (no hierarchy)
amova_result <- poppr.amova(genind_data, ~Population, within = FALSE)

# Test significance with 1000 permutations
amova_signif <- randtest(amova_result, nrepet = 1000)
plot(amova_signif)  # Check p-value distribution
amova_signif

# Scale genotypes
genind_scaled <- scaleGen(genind_data, NA.method = "mean")

# Run PCA
pca <- dudi.pca(genind_scaled, cent = FALSE, scale = FALSE, scannf = FALSE)

# Plot
library(ggplot2)
ggplot(data.frame(pca$li, Population = pop(genind_data)), 
       aes(x = Axis1, y = Axis2, color = Population)) +
  geom_point(size = 3) +
  stat_ellipse() +  # Optional: Add ellipses
  theme_minimal()
dapc_result <- dapc(genind_data, n.pca = 10, n.da = 2)  # Adjust n.pca as needed
scatter(dapc_result, col = rainbow(3))



##creating heatmap
library(ggplot2)
library(pheatmap)
library(reshape2)

fst_matrix <- matrix(c(
  NA, 0.0190, 0.0094, -0.0038, 0.507, 0.505, 0.489, 0.509, 0.614,
  0.0190, NA, 0.0027, 0.0200, 0.550, 0.548, 0.591, 0.577, 0.686,
  0.0094, 0.0027, NA, 0.0130, 0.499, 0.498, 0.513, 0.502, 0.634,
  -0.0038, 0.0200, 0.0130, NA, 0.518, 0.516, 0.527, 0.522, 0.649,
  0.507, 0.550, 0.499, 0.518, NA, 0.0076, 0.192, 0.031, 0.488,
  0.505, 0.548, 0.498, 0.516, 0.0076, NA, 0.191, 0.021, 0.484,
  0.489, 0.591, 0.513, 0.527, 0.192, 0.191, NA, 0.193, 0.135,
  0.509, 0.577, 0.502, 0.522, 0.031, 0.021, 0.193, NA, 0.445,
  0.614, 0.686, 0.634, 0.649, 0.488, 0.484, 0.135, 0.445, NA
), nrow = 9, byrow = TRUE)
rownames(fst_matrix) <- c("W3", "W1", "W", "W4", "Q1", "Q2", "Q3", "Q4", "Q5")
colnames(fst_matrix) <- c("W3", "W1", "W2", "W4", "Q1", "Q2", "Q3", "Q4", "Q5")

pheatmap(
  fst_matrix,
  color = colorRampPalette(c("blue", "white", "red"))(100),  # Blue (low) -> White (0) -> Red (high)
  breaks = seq(-0.01, 0.7, length.out = 101),              # Scale from -0.01 to 0.7
  cluster_rows = TRUE,                                      # Cluster rows
  cluster_cols = TRUE,                                     # Cluster columns
  display_numbers = TRUE,                                  # Show FST values
  number_format = "%.3f",                                  # 3 decimal places
  main = "Pairwise FST Heatmap"
)

#####Treemix
#### in bash
module load Stacks
populations -V ./final.recode.vcf - O ./ -M ./popmap.txt â€“treemix
gzip tree_final.treemix
module load Miniconda3
for i in {1..8};
do
treemix -i tree_final.treemix.gz -m ${i} -bootstrap -k 500 -o BMSB_${i} > treemix_${i}_log 
done

########## in R
source("plotting_funcs.R") 
plot_residuals("BMSB_7") 
plot_tree("BMSB_7")
snp_vcf2 = read.vcfR("final.recode.vcf")
vcf <- read.vcfR("final.recode.vcf")



